### ISSUE TITLE
Add support for optparse's 'float' and 'complex' types. - [merged]

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#894]
In GitLab by @ymdatta on Oct 28, 2018, 05:08

_Merges master -> master_

This helps normalize_from_setuptools of manager.py
to handle all optparse's standard-option-types.

Fixes #452

**Comment by asottile** [comment]
In GitLab by @asottile on Oct 28, 2018, 08:09

Commented on [src/flake8/options/manager.py line 162](https://github.com/PyCQA/flake8/compare/4773995430e56f5669861438cc3ce2845953ad51..cc20e35058da474bd56eea7cf893a872506926e6#diff-4dbff2b18da772d0cf3dc7898f4a4e45R162)



mind adding a quick test demonstrating this change?

**Comment by asottile** [comment]
In GitLab by @ymdatta on Oct 28, 2018, 09:49

added 1 commit

<ul><li>ccd9beb2 - Added tests for optparse&#39;s &#39;float&#39; and &#39;complex&#39; types</li></ul>

[Compare with previous version](https://gitlab.com/pycqa/flake8/merge_requests/261/diffs?diff_id=26060681&start_sha=6341d1382e0ab14043396e8a9fbb18c27a3e236c)

**Comment by asottile** [comment]
In GitLab by @codecov on Oct 28, 2018, 09:50

# [Codecov](https://codecov.io/gl/pycqa/flake8/pull/261?src=pr&el=h1) Report
> Merging [#261](https://codecov.io/gl/pycqa/flake8/pull/261?src=pr&el=desc) into [master](https://codecov.io/gl/pycqa/flake8/commit/4773995430e56f5669861438cc3ce2845953ad51?src=pr&el=desc) will **increase** coverage by `0.04%`.
> The diff coverage is `75%`.

[![Impacted file tree graph](https://codecov.io/gl/pycqa/flake8/pull/261/graphs/tree.svg?width=650&token=QL6TVnaXih&height=150&src=pr)](https://codecov.io/gl/pycqa/flake8/pull/261?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master     #261      +/-   ##
==========================================
+ Coverage   70.79%   70.84%   +0.04%     
==========================================
  Files          27       27              
  Lines        2349     2353       +4     
  Branches      384      386       +2     
==========================================
+ Hits         1663     1667       +4     
+ Misses        608      606       -2     
- Partials       78       80       +2
```


| [Impacted Files](https://codecov.io/gl/pycqa/flake8/pull/261?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [src/flake8/options/manager.py](https://codecov.io/gl/pycqa/flake8/pull/261/diff?src=pr&el=tree#diff-c3JjL2ZsYWtlOC9vcHRpb25zL21hbmFnZXIucHk=) | `90.9% <75%> (+0.28%)` | :arrow_up: |

------

[Continue to review full report at Codecov](https://codecov.io/gl/pycqa/flake8/pull/261?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gl/pycqa/flake8/pull/261?src=pr&el=footer). Last update [4773995...cc20e35](https://codecov.io/gl/pycqa/flake8/pull/261?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Nov 1, 2018, 05:40

Commented on [tests/unit/test_option.py line 26](https://github.com/PyCQA/flake8/compare/4773995430e56f5669861438cc3ce2845953ad51..ccd9beb26dac68ada62d3c5fd58c0e4ffa3d79c3#diff-f53fc80a181ee2235ae5ae0e7e709828R26)



I think this test as written isn't exercising what it seems you thought it was exercising. The code above checks the `type` value passed to the `Option()` class. This test doesn't seem to ever have a type specified that would exercise those branches. What do you think of re-writing this to be:

```py
@pytest.mark.parametrize('opttype,str_val,expected', [
    ('float', '2', 2.0),
    ('complex', '2', (2+0j)),
])
def test_to_support_optparses_standard_types(opttype, str_val, expected):
    """<Your docstring here>"""
    opt = manager.Option('-t', '--test', type=opttype)
    assert opt.normalize_from_setuptools(str_val) == expected
```

**Comment by asottile** [comment]
In GitLab by @ymdatta on Nov 5, 2018, 08:08

@sigmavirus24 : you are right, i am sorry and i understood my mistake.

So i made the changes and ran the `test_option.py` using pytest, but i was getting errors, errors being from the tests we wrote. So, i used gdb to find out why i was getting the error, as our `normalize_from_setuptools()` should return correct results, it turned out that the changes in my manager.py aren't reflected when the test_option.py calls manager.py.

TL;DR: getting erros in tests, because test doesn't recognize changes made in the manager.py

I am running tests with the command: `pytest-3 test_option.py`

I also checked documentation to see if i am missing something, but there was no information on how to run tests.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Nov 6, 2018, 05:32

> I also checked documentation to see if i am missing something, but there was no information on how to run tests.

There should be docs around using `tox` to run the tests. If there aren't docs around that, I'd welcome a separate PR that documents that.

**Comment by asottile** [comment]
In GitLab by @ymdatta on Nov 8, 2018, 19:12

Commented on [tests/unit/test_option.py line 26](https://github.com/PyCQA/flake8/compare/4773995430e56f5669861438cc3ce2845953ad51..ccd9beb26dac68ada62d3c5fd58c0e4ffa3d79c3#diff-f53fc80a181ee2235ae5ae0e7e709828R26)



changed this line in [version 5 of the diff](https://gitlab.com/pycqa/flake8/merge_requests/261/diffs?diff_id=26929385&start_sha=ccd9beb26dac68ada62d3c5fd58c0e4ffa3d79c3#2ae10c377ae1834f9f53e35ae0260c2d40c0f7f6_26_26)

**Comment by asottile** [comment]
In GitLab by @ymdatta on Nov 8, 2018, 19:12

added 1 commit

<ul><li>3b161305 - test_option:Modify the tests to check support for optparse&#39;s types.</li></ul>

[Compare with previous version](https://gitlab.com/pycqa/flake8/merge_requests/261/diffs?diff_id=26929385&start_sha=ccd9beb26dac68ada62d3c5fd58c0e4ffa3d79c3)

**Comment by asottile** [comment]
In GitLab by @asottile on Jan 3, 2019, 17:16

@ymdatta can you fix the linting error -- then I can merge this :thumbsup: 

```diff
--- a/tests/unit/test_option.py
+++ b/tests/unit/test_option.py
@@ -25,7 +25,7 @@ def test_to_optparse():
 
 @pytest.mark.parametrize('opttype,str_val,expected', [
     ('float', '2', 2.0),
-    ('complex', '2', (2+0j)),
+    ('complex', '2', (2 + 0j)),
 ])
 def test_to_support_optparses_standard_types(opttype, str_val, expected):
     """Show that optparse converts float and complex types correctly."""
```

**Comment by asottile** [comment]
In GitLab by @asottile on Jan 3, 2019, 17:16

approved this merge request

**Comment by asottile** [comment]
In GitLab by @ymdatta on Jan 3, 2019, 21:03

added 1 commit

<ul><li>cc20e350 - Fixed linting errors</li></ul>

[Compare with previous version](https://gitlab.com/pycqa/flake8/merge_requests/261/diffs?diff_id=30681677&start_sha=3b161305003448f67e5f9c4847ff196e8d2ac716)

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jan 17, 2019, 05:50

resolved all discussions

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jan 17, 2019, 05:50

mentioned in commit 0f3f4ff9b4e7901695484ee1ea1fbcbc19632149

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jan 17, 2019, 05:50

merged
